transaction/std/src/memo_builder/burn_redemption_memo_builder.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+
+//! Defines the BurnRedemptionMemoBuilder.
+//! This MemoBuilder policy implements Burn Redemption tracking usin memos, as
+//! envisioned in MCIP #TODO.
+
+use super::{
+    memo::{BurnRedemptionMemo, DestinationMemo, DestinationMemoError, UnusedMemo},
+    MemoBuilder,
+};
+use crate::ChangeDestination;
+use mc_account_keys::{burn_address, PublicAddress, ShortAddressHash};
+use mc_transaction_core::{tokens::Mob, MemoContext, MemoPayload, NewMemoError, Token};
+
+/// This memo builder attaches 0x0001 Burn Redemption Memos to an output going
+/// to the designated burn address, and 0x0200 Destination Memos to change
+/// outputs. Only a single non-change output is allowed, and it must go to the
+/// designated burn address.
+///
+/// Usage:
+/// You should usually use this like:
+///
+///   let memo_data = [1; 64]; // TODO: contents of this are still not designed.
+///   let mut mb = BurnRedemptionMemoBuilder::new(memo_data);
+///   mb.enable_destination_memo();
+///
+/// Then use it to construct a transaction builder.
+///
+/// A memo builder configured this way will use 0x0001 Burn Redemption Memo
+/// on the burn output and 0x0200 Destination Memo on the change output.
+///
+/// If mb.enable_destination_memo() is not called 0x0000 Unused will appear on
+/// change output, instead of 0x0200 Destination Memo.
+///
+/// When invoking the transaction builder, the change output must be created
+/// last. If the burn output is created after the change output, an error will
+/// occur.
+///
+/// If more than one burn output is created, an error will be returned.
+#[derive(Clone, Debug)]
+pub struct BurnRedemptionMemoBuilder {
+    // The memo data we will attach to the burn output.
+    memo_data: [u8; 64],
+    // Whether destination memos are enabled.
+    destination_memo_enabled: bool,
+    // Tracks if we already wrote a burn memo, for error reporting
+    wrote_burn_memo: bool,
+    // Tracks if we already wrote a destination memo, for error reporting
+    wrote_destination_memo: bool,
+    // Tracks the amount being burn
+    burn_amount: u64,
+    // Tracks the fee
+    fee: u64,
+}
+
+impl BurnRedemptionMemoBuilder {
+    /// Construct a new BurnRedemptionMemoBuilder.
+    pub fn new(memo_data: [u8; 64]) -> Self {
+        Self {
+            memo_data,
+            destination_memo_enabled: false,
+            wrote_burn_memo: false,
+            wrote_destination_memo: false,
+            burn_amount: 0,
+            fee: Mob::MINIMUM_FEE,
+        }
+    }
+    /// Enable destination memos
```

Comment:
This would deviate from other memo builders. Personally I am also not a fan of methods that accept a bunch of boolean/integer parameters since it requires looking at the signature to see what each of them means.

Reply:
That's fair - but usually in builder formats the convention is to return the object so it can be built 
```rust
        ChannelBuilder::new(env)
            .keepalive_permit_without_calls(true)
            .keepalive_time(Duration::from_secs(1))
```

So perhaps we might be able to do something like:
```rust
let memo_builder = BurnRedemptionMemoBuilder::new([2u8; 64]).enable_destination_memo().set_fee(4000)?
```

Reply:
This is a bit different than `ChannelBuilder` due to `ChannelBuilder` being used to construct a single `Channel`, at which point `self` is consumed. The memo builder is being called repeatedly.
In the current state of things, I believe we will run into issues if we made `set_fee` be `fn set_fee(mut self, ...) -> Self` (which is what `ChannelBuilder` does to achieve this call chaining), since `TransactionBuilder` stores a `Box<dyn MemoBuilder>` and this will "unbox" it

Reply:
yeah it's not the same kind of chaining builder, the transaction builder doesn't work that way either

Reply:
Fair enough!